### PR TITLE
OHRM5X-1121: Improve selection of current menu item

### PR DESCRIPTION
**Pull request by RajithaKumara**

### CODE REVIEW

**Comment by RajithaKumara** [comment]
Revert this [line](https://github.com/orangehrm/orangehrm/pull/1148/files#diff-2ed7b0d4466cb521f3b746baa69f076e6af0c0bbd83e5801ecaa26e78ddb8175R197) to fix OHRM5X-1176

File: symfony/plugins/orangehrmCorePlugin/Menu/DetailedMenuItem.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+{
+    private int $id;
+    private string $menuTitle;
+    private ?array $additionalParams = [];
+    private ?string $module = null;
+    private ?string $screen = null;
+    /**
+     * @var DetailedMenuItem[]
+     */
+    private array $childMenuItems = [];
```
Comment (devishke-orange) [inline symfony/plugins/orangehrmCorePlugin/Menu/DetailedMenuItem.php:35]: maybe keep a blank line before doc comment
```suggestion
    private ?string $screen = null;
    
    /**
     * @var DetailedMenuItem[]
     */
    private array $childMenuItems = [];
```

File: symfony/plugins/orangehrmCorePlugin/Service/MenuService.php [hunk 2]
```diff
@@ -68,291 +59,210 @@ public function getMenuDao(): MenuDao
...
     /**
      * @param string $moduleName
      * @param array $menuTitles
      * @return int
-     * @throws DaoException
      */
     public function enableModuleMenuItems(string $moduleName, array $menuTitles = []): int
     {
         return $this->getMenuDao()->enableModuleMenuItems($moduleName, $menuTitles);
     }
```
Comment (devishke-orange) [inline symfony/plugins/orangehrmCorePlugin/Service/MenuService.php:80]: like a wrapper function can it be removed?

File: symfony/plugins/orangehrmCorePlugin/test/Service/MenuServiceTest.php [hunk 3]
```diff
@@ -21,17 +21,14 @@
...
 use OrangeHRM\Tests\Util\TestCase;
 use OrangeHRM\Tests\Util\TestDataService;
 
 /**
- * Description of MenuServiceTest
  * @group Core
  * @group Service
  */
-class MenuServiceTest extends TestCase
+class MenuServiceTest // extends TestCase
```
Comment (devishke-orange) [inline symfony/plugins/orangehrmCorePlugin/test/Service/MenuServiceTest.php:31]: will the tests be written later?

**Review by devishke-orange — CHANGES_REQUESTED** [review]
No issues with formatting as far as I could see 👍 . Have a few suggestions and some questions

File: symfony/plugins/orangehrmCorePlugin/test/Service/MenuServiceTest.php (hunk 3 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/test/Service/MenuServiceTest.php:31]: yes. need to address

File: symfony/plugins/orangehrmCorePlugin/Service/MenuService.php (hunk 2 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Service/MenuService.php:80]: yes. will try to attempt it. used multiple places

File: installer/upgrader/Migrations/V5/AttendanceHelper.php [hunk 4]
```diff
@@ -96,7 +96,7 @@ public function updatePunchOutTimezoneOffset(string $offset, string $timezone):
     {
         $q = $this->createQueryBuilder();
         $q->update('ohrm_attendance_record')
-            ->set('ohrm_attendance_record.punch_out_timezone', ':punchOutTimezone')
+            ->set('ohrm_attendance_record.punch_in_timezone_name', ':punchOutTimezone')
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply installer/upgrader/Migrations/V5/AttendanceHelper.php:99]: nice catch

File: installer/dbscript/dbscript-2.sql [hunk 5]
```diff
@@ -4477,3 +4477,112 @@ SET @admin_role_id := (SELECT `id` FROM ohrm_user_role WHERE `name` = 'Admin' LI
...
+(@ess_role_id, @editAttendanceRecordScreenId, 1, 0, 0, 0),
+(@admin_role_id, @editEmployeeAttendanceRecordScreenId, 1, 0, 0, 0),
+(@supervisor_role_id, @editEmployeeAttendanceRecordScreenId, 1, 0, 0, 0),
+(@admin_role_id, @proxyPunchInPunchOutScreenId, 1, 0, 0, 0),
+(@supervisor_role_id, @proxyPunchInPunchOutScreenId, 1, 0, 0, 0);
+
+SET @time_menu_item := (SELECT `id` FROM ohrm_menu_item WHERE menu_title = 'Time' LIMIT 1);
+UPDATE `ohrm_menu_item` SET `status` = '0' WHERE `parent_id` = @time_menu_item;
+
+DROP TABLE `ohrm_data_group_screen`;
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply installer/dbscript/dbscript-2.sql:4588]: not using it. so going to drop it

File: symfony/plugins/orangehrmAdminPlugin/Menu/QualificationsMenuConfigurator.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+class QualificationsMenuConfigurator implements MenuConfigurator
+{
+    use ModuleScreenHelperTrait;
+
+    /**
+     * @inheritDoc
+     */
+    public function configure(Screen $screen): ?MenuItem
+    {
+        $this->getCurrentModuleAndScreen()->overrideScreen(str_replace('view', 'save', $screen->getActionUrl()));
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply symfony/plugins/orangehrmAdminPlugin/Menu/QualificationsMenuConfigurator.php:36]: common pattern there in those screens, e.g. `saveSkill` => `viewSkill`

File: symfony/plugins/orangehrmAttendancePlugin/Menu/AttendanceMenuConfigurator.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+class AttendanceMenuConfigurator implements MenuConfigurator
+{
+    /**
+     * @inheritDoc
+     */
+    public function configure(Screen $screen): ?MenuItem
+    {
+        $moduleScreen = ModuleScreenHelper::getCurrentModuleAndScreen();
+        $moduleScreen->overrideModule('time');
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply symfony/plugins/orangehrmAttendancePlugin/Menu/AttendanceMenuConfigurator.php:35]: :ok_hand:

**Review by devishke-orange — CHANGES_REQUESTED** [review]
Looks good 👍 I tried going through the refactor code but having a little trouble understanding what exactly is happening 😅

**Comment by RajithaKumara** [comment]
> Looks good +1 I tried going through the refactor code but having a little trouble understanding what exactly is happening sweat_smile

Not only refactoring. added menu item caching support :relaxed:

File: symfony/plugins/orangehrmCorePlugin/test/fixtures/MenuService.yaml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  - { id: 2, name: 'Development', country_code: 'US' }
+  - { id: 3, name: 'Support', country_code: 'LK' }
+
+EmpLocations:
+  - { emp_number: 1, location_id: 1 }
+  - { emp_number: 2, location_id: 2 }
+  - { emp_number: 3, location_id: 2 }
+  - { emp_number: 4, location_id: 2 }
+  - { emp_number: 5, location_id: 3 }
+
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/test/fixtures/MenuService.yaml:60]: This fixture depends on `symfony/test/phpunit/fixtures/MenuItem.yaml`. it contain set of data 
https://github.com/orangehrm/orangehrm/pull/1166/files#diff-25d276e3ff97cad45a3958ecaf5c4e70861889251e444944af56d663e9a15c0a

File: symfony/plugins/orangehrmCorePlugin/Service/MenuService.php (hunk 2 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Service/MenuService.php]: this is removed and again added to reset menu cache

File: symfony/plugins/orangehrmCorePlugin/test/fixtures/MenuService.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (devishke-orange) [reply symfony/plugins/orangehrmCorePlugin/test/fixtures/MenuService.yaml:63]: Ohh okay 👍

**Review by devishke-orange — COMMENTED** [review]
Looks good. No issues that I came across 👍

**Pull request merged**